go/apps/api/routes/v2_keys_verify_key/handler.go

```diff
@@ -0,0 +1,261 @@
+	if key.Key.WorkspaceID != auth.AuthorizedWorkspaceID {
+		return s.JSON(http.StatusOK, res)
+	}
+
+	// Check if API is deleted
+	if key.Key.ApiDeletedAtM.Valid {
+		return s.JSON(http.StatusOK, res)
+	}
+
+	// FIXME: We are leaking a keys existance here... by telling the user that he doesn't have perms
+	err = auth.Verify(ctx, keys.WithPermissions(rbac.Or(
+		rbac.T(rbac.Tuple{
+			ResourceType: rbac.Api,
+			ResourceID:   "*",
+			Action:       rbac.VerifyKey,
+		}),
+		rbac.T(rbac.Tuple{
+			ResourceType: rbac.Api,
+			ResourceID:   key.Key.ApiID,
+			Action:       rbac.VerifyKey,
+		}),
+	)))
+	if err != nil {
+		return err
+	}
+
+	opts := []keys.VerifyOption{keys.WithIPWhitelist(), keys.WithApiID(req.ApiId), keys.WithTags(ptr.SafeDeref(req.Tags))}
+
+	// If a custom cost was specified, use it, otherwise use a DefaultCost of 1
+	if req.Credits != nil {
+		opts = append(opts, keys.WithCredits(req.Credits.Cost))
+	} else if key.Key.RemainingRequests.Valid {
+		opts = append(opts, keys.WithCredits(DefaultCost))
+	}
+
+	if req.Ratelimits != nil {
+		opts = append(opts, keys.WithRateLimits(*req.Ratelimits))
+	} else {
+		// check auto applied ratelimits
+		opts = append(opts, keys.WithRateLimits(nil))
+	}
+
+	if req.Permissions != nil {
+		query, queryErr := convertPermissionsToQuery(*req.Permissions)
+		if queryErr != nil {
+			return queryErr
+		}
+
+		opts = append(opts, keys.WithPermissions(query))
+	}
+
+	err = key.Verify(ctx, opts...)
+	if err != nil {
+		return err
+	}
+
+	res.Data = openapi.KeysVerifyKeyResponseData{
+		Code:        key.ToOpenAPIStatus(),
+		Valid:       key.Valid,
+		Enabled:     ptr.P(key.Key.Enabled),
+		Name:        ptr.P(key.Key.Name.String),
+		Permissions: ptr.P(key.Permissions),
+		Roles:       ptr.P(key.Roles),
+		KeyId:       ptr.P(key.Key.ID),
+
+		Credits:    nil,
+		Expires:    nil,
+		Identity:   nil,
+		Meta:       nil,
+		Ratelimits: nil,
+	}
+
+	remaining := key.Key.RemainingRequests
+	if remaining.Valid {
+		res.Data.Credits = ptr.P(remaining.Int32)
+	}
+
+	if key.Key.Expires.Valid {
+		res.Data.Expires = ptr.P(key.Key.Expires.Time.UnixMilli())
+	}
+
+	if key.Key.Meta.Valid {
+		err = json.Unmarshal([]byte(key.Key.Meta.String), &res.Data.Meta)
+		if err != nil {
+			return fault.Wrap(err, fault.Code(codes.App.Internal.UnexpectedError.URN()),
+				fault.Internal("unable to unmarshal key meta"),
+				fault.Public("We encountered an error while trying to unmarshal the key meta data."),
+			)
+		}
+	}
+
+	if key.Key.IdentityID.Valid {
+		res.Data.Identity = &openapi.Identity{
+			ExternalId: key.Key.ExternalID.String,
+			Id:         key.Key.IdentityID.String,
+			Ratelimits: nil,
+			Meta:       nil,
+		}
+
+		for _, ratelimit := range key.GetRatelimitConfigs() {
+			if ratelimit.IdentityID == "" {
+				continue
+			}
+
+			res.Data.Identity.Ratelimits = append(res.Data.Identity.Ratelimits, openapi.RatelimitResponse{
+				AutoApply: ratelimit.AutoApply == 1,
+				Duration:  int64(ratelimit.Duration),
```

Comment:
can you double chech this? if `ratelimit.Duration` is a `time.Duration` then just casting it would return the nanosecond representation, we need to call `ratelimit.Duration.UnixMilli()`

Reply:
its a normal int